distributed/protocol/serialize.py

```diff
@@ -461,15 +469,27 @@ def merge_and_deserialize(header, frames, deserializers=None):
     deserialize
     serialize_and_split
     """
-    merged_frames = []
     if "split-num-sub-frames" not in header:
         merged_frames = frames
     else:
+        merged_frames = []
         for n, offset in zip(header["split-num-sub-frames"], header["split-offsets"]):
-            if n == 1:
-                merged_frames.append(frames[offset])
+            if n == 0:
+                merged = bytearray(0)
+            elif n == 1:
+                merged = frames[offset]
             else:
-                merged_frames.append(bytearray().join(frames[offset : offset + n]))
+                subframes = frames[offset : offset + n]
+                first = subframes[0]
+                if isinstance(first, memoryview) and first.contiguous:
+                    try:
+                        merged = merge_memoryviews(subframes)
+                    except ValueError:
+                        merged = bytearray().join(subframes)
+                else:
+                    merged = bytearray().join(subframes)
```

Comment:
The whole block 477:490 could be replaced by
```python
subframes = frames[offset : offset + n]
try:
    merged = merge_memoryviews(subframes)
except (ValueError, TypeError):
    merged = bytearray().join(subframes)
```
after small changes to merge_memoryviews shown below.